engine/access/rpc/engine.go

```diff
@@ -298,24 +313,25 @@ func (e *Engine) serveUnsecureGRPC() {
 	e.addrLock.Lock()
 	e.unsecureGrpcAddress = l.Addr()
 	e.addrLock.Unlock()
-
 	e.log.Debug().Str("unsecure_grpc_address", e.unsecureGrpcAddress.String()).Msg("listening on port")
+	ready()
 
 	err = e.unsecureGrpcServer.Serve(l) // blocking call
 	if err != nil {
-		e.log.Fatal().Err(err).Msg("fatal error in unsecure grpc server")
+		e.log.Err(err).Msg("fatal error in unsecure grpc server")
+		ctx.Throw(err)
 	}
 }
 
-// serveSecureGRPC starts the secure gRPC server
-// When this function returns, the server is considered ready.
-func (e *Engine) serveSecureGRPC() {
-
+// serveSecureGRPCWorker is a worker routine which starts the secure gRPC server.
+// The ready callback is called after the server address is bound and set.
+func (e *Engine) serveSecureGRPCWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
 	e.log.Info().Str("secure_grpc_address", e.config.SecureGRPCListenAddr).Msg("starting grpc server on address")
 
 	l, err := net.Listen("tcp", e.config.SecureGRPCListenAddr)
 	if err != nil {
 		e.log.Err(err).Msg("failed to start the grpc server")
+		ctx.Throw(err)
```

Comment:
(technically not a part of this PR)
I'm a bit confused by this pattern.

the add worker accepts the `type ComponentWorker func(ctx irrecoverable.SignalerContext, ready ReadyFunc)`. which means the implementer must know `irrecoverable.SignalerContext` and to call ctx.Throw on error. but that seems unnecessary. This could be changed to `type ComponentWorker func(ctx context.Context, ready ReadyFunc) err` and the throw could happen outside in case of error.

Reply:
Each `ComponentWorker` is [started as a goroutine](https://github.com/onflow/flow-go/blob/60cc312acee146f49ebb62bb9a9314908689be95/module/component/component.go#L252-L260), so they can't directly return an error right now. 

You're right though, we could have a wrapper function around the worker, which just receives any error and throws it. I don't really see the downside to implementors needing to know about `SignalerContext` though -- they're implementing an interface for a worker, they need to know how that interface works. Using the types defined by the interface is part of that contract. Whether `SignalerContext` is a useful construct in general is a different question...

There is some benefit for errors which directly cause a crash to be differentiated from other errors, which the `SignalerContext` achieves. Also, the intention is for `SignalerContext` to be passed down beyond the worker's stack frame (as the context in functions accepting a context), otherwise there is little use for it. On the other hand, we rarely do this in practice (rather, we return exceptions up the stack to the top-level worker function, which throws). Another possible use-case is to enable worker routines to pass the `SignalerContext` to a child goroutine, so it can respond to exceptions properly, though we generally avoid doing this too.

Overall I'm in favour of the current signature, because it differentiates returning an error (maybe benign or maybe an exception) from throwing an exception in the business logic. However, I agree that some of the original ideas for how `SignalerContext`s would be used aren't being applied in practice.